### ISSUE TITLE
🔌 Plugin: Prometheus metrics + analytics plugin

### ISSUE THREAD

**Issue by aspic** [backstage/backstage#14505]
### 🔖 Summary

Hey, and thanks for this great open source project!

We've been developing two simple plugins in-house:

1. A metrics plugin, exposing Prometheus metrics for Backstage and means for collecting arbitrary metrics 
2. An analytics plugin, feeding tracking data (views) to Backstage back-end and exposing this as metrics

```
# HELP app_name_page_views Counts page views
# TYPE app_name_page_views counter
app_name_page_views{page="/api-docs"} 2
app_name_page_views{page="/catalog/default/Group/some-team"} 1
app_name_page_views{page="/"} 5
```

We can then visualize this with the Prometheus visualizer from Backstage.

**Question: Is this something you'd be interested in us open-sourcing / contributing back to backstage?**

### 👀 Have you spent some time to check if this plugin request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Contributing Guidelines](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by awanlin** [comment]
Good ideas here! Just have a few questions:

For item 1, would this be an actual plugin or just adding more metrics like those that currently exists for the catalog and scaffolder? https://github.com/backstage/backstage/blob/master/contrib/docs/tutorials/prometheus-metrics.md

For item 2, would you use the existing Analytics API that's collecting some of this data already? If we didn't want to view this in Backstage but in a tool like Grafana I assume the metrics would still be available for that?

**Comment by jhaals** [comment]
Hey!

Thanks for the suggestion, metrics is definitely on our roadmap. We are currently waiting on OpenTelemetry to release a stable metrics API. It's currently in RC stage and status can be seen here https://github.com/open-telemetry/opentelemetry-js#feature-status. Once that is done we aim to incorporate both metrics and tracing into Backstage backends. I don't think this necessarily need to be a traditional plugin, rather APIs that you have available as a plugin developer to instrument your plugin. OpenTelemetry also come with an exporter for Prometheus.

**Comment by aspic** [comment]
Thanks for chiming in here! Swear I searched for Prometheus in the docs before implementing it as a plugin, but that tutorial there flew under my radar.

@awanlin To answer your question, it's implemented as a plugin now, injected with a Prometheus client. So that the client can be passed around to other plugins, and metrics will be exposed on the configured route from the plugin. The tutorial probably renders the plugin redundant, but would be nice if it was searchable from the Backstage docs! :-)

Yup! So the frontend part is implementing the Analytics API and passing events to an analytics endpoint, then that event increments one or multiple Prometheus counters. Yeah, can probably build much nicer graphs for it in Grafana! Since tracking is for internal use only, it's super nice avoiding sending data to the tracking companies.

@jhaals ah, that's good to know! Looks like that would at least render a prometheus metrics plugin redundant? Please let me know if any other parts of this Issue is interesting (ie. an analytics plugin for Prometheus).

**Comment by jhaals** [comment]
> @jhaals ah, that's good to know! Looks like that would at least render a prometheus metrics plugin redundant? Please let me know if any other parts of this Issue is interesting (ie. an analytics plugin for Prometheus).

Yeah I think a plugin for the purpose of exposing a scraping endpoint for prometheus would be redundant. There might still be room for a plugin for working with or displaying data that's already in Prometheus but I guess there's a balance not to reinvent Grafana 😅

**Comment by aspic** [comment]
Agree, thanks for talking this through with me! :-)

**Comment by awanlin** [comment]
@aspic, I do think the 2nd item would have a lot of value for those in the community who can't use Google Analytics due to legal concerns. I've often considered doing this but just logging the results to standard out so we could maybe use that with Elasticsearch and Grafana. Maybe you could log a new issue for just that?

**Comment by aspic** [comment]
@awanlin will do! I guess more and more people are in that boat, and it's very nice to avoid the "roundtrip" to google just to get some basic usage metrics!

**Comment by awanlin** [comment]
@aspic, awesome! Tag me in it so I can keep track 🚀
